packages/core/src/storage/compare.ts

```diff
@@ -108,25 +177,31 @@ export class StorageLayoutComparator {
     const layoutChange = this.getLayoutChange(original, updated);
 
     if (updated.retypedFrom && layoutChange) {
-      return { kind: 'layoutchange', original, updated, change: layoutChange };
+      return { kind: 'layoutchange', original, updated, change: layoutChange, cost: LAYOUTCHANGE_COST };
+    } else if (nameChange && endMatchesGap(original, updated)) {
+      // original is a gap that was consumed by a non-gap ending at the same slot, which is safe
+      return { kind: 'finishgap', original, updated, cost: FINISHGAP_COST };
+    } else if (typeChange && typeChange.kind === 'array shrink' && endMatchesGap(original, updated)) {
+      // original and updated are a gap that was shrunk and ends at the same slot, which is safe
+      return { kind: 'shrinkgap', change: typeChange, original, updated, cost: SHRINKGAP_COST };
     } else if (typeChange && nameChange) {
       return { kind: 'replace', original, updated };
     } else if (nameChange) {
       return { kind: 'rename', original, updated };
     } else if (typeChange) {
-      return { kind: 'typechange', change: typeChange, original, updated };
+      return { kind: 'typechange', change: typeChange, original, updated, cost: TYPECHANGE_COST };
     } else if (layoutChange && !layoutChange.uncertain) {
       // Any layout change should be caught earlier as a type change, but we
       // add this check as a safety fallback.
-      return { kind: 'layoutchange', original, updated, change: layoutChange };
+      return { kind: 'layoutchange', original, updated, change: layoutChange, cost: LAYOUTCHANGE_COST };
     }
   }
 
   getLayoutChange(original: StorageField, updated: StorageField): LayoutChange | undefined {
     const validPair = ['uint8', 'bool'];
     const knownCompatibleTypes =
       validPair.includes(original.type.item.label) && validPair.includes(updated.type.item.label);
```

Comment:
I know that is not part of this PR, but do we really think there is only one valid pair, and that they are always pairs, we'll ever consider?

I would have done 
```suggestion
    const validChanges = [['uint8', 'bool']];
    const knownCompatibleTypes = validChanges.some(group => goup.includes(original.type.item.label) && group.includes(updated.type.item.label));
```

Also, that means a bool can become a uint8, but also that a uint8 can become a bool. Is that ok ?

Reply:
Reworked this part of the code to include this suggestion and also ensure that it does not omit layout changes for these valid pairs.